### ISSUE TITLE
Change NVIDIA dataset iterator to proper `DataLoader`

### ISSUE THREAD

**Issue by gpauloski** [gpauloski/llm-pytorch#2]
Currently, we use an iterator that wraps a bunch of `DataLoader`s for each shard, but this means we cannot pass a single `DataLoader` to `colossalai.initialize` which drops last batches based on the gradient accumulation. A single `DataLoader` and `Sampler` would also make state save/resume easier.
